libglusterfs/src/glusterfs/gf-buf.h

```diff
@@ -0,0 +1,160 @@
+#ifndef _GF_BUF_H_
+#define _GF_BUF_H_
+
+#include "glusterfs/common-utils.h"
+
+#ifndef GF_BUF_STACK_THRESHOLD
+#define GF_BUF_STACK_THRESHOLD 64
+#endif
+
+typedef struct gf_buf {
+    unsigned long heap : 1;
+    unsigned long allocated : __BITS_PER_LONG - 1;
+    unsigned long used;
+    char *data;
+} gf_buf_t;
+
+void
+gf_buf_cleanup(struct gf_buf *);
+
+#define gf_buf_decl_null(var)                                                  \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {0, 0, 0, NULL}
+
+#define gf_buf_decl_empty(var)                                                 \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {0, 0, 1, ""}
+
+#define gf_buf_decl_string(var, str)                                           \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {              \
+        0, 0, strlen(str) + 1, (str)}
+
+#define __gf_buf_malloc(size) GF_MALLOC((size), gf_common_mt_char)
+
+#define __gf_buf_free(ptr) GF_FREE((ptr))
+
+#define __gf_buf_assign(buf, str, size)                                        \
+    do {                                                                       \
+        if (size > GF_BUF_STACK_THRESHOLD) {                                   \
+            (buf)->data = memcpy(__gf_buf_malloc(size), str, size);            \
+            (buf)->heap = 1;                                                   \
+        } else {                                                               \
+            (buf)->data = memcpy(alloca(size), str, size);                     \
+            (buf)->heap = 0;                                                   \
+        }                                                                      \
+    } while (0)
+
+#define gf_buf_assign(buf, str)                                                \
+    do {                                                                       \
+        size_t __sz = strlen(str) + 1;                                         \
```

Comment:
To make it more efficient, sizes should be rounded to next multiple of 16 for stack buffers and 64/128 for heap buffers.

Reply:
Why not 16 in both cases? IIUC this is enough if we want to help the compiler to use SIMD stuff on modern 64-bit hardware.

Reply:
> Why not 16 in both cases? IIUC this is enough if we want to help the compiler to use SIMD stuff on modern 64-bit hardware.

For cache line alignment, I reckon.